From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: zhangzujian <zhangzujian.7@gmail.com>
Date: Sat, 20 Sep 2026 00:00:00 +0000
Subject: [PATCH] use L4 tuple hash for policy route ECMP

Hash policy ECMP the same way as static-route selection_fields
tp_src/tp_dst: TCP/UDP/SCTP include L4 ports, other IP traffic
hashes on addresses and ip_proto.

Signed-off-by: zhangzujian <zhangzujian.7@gmail.com>
---
 northd/northd.c | 46 +40 -6
 1 file changed, 40 insertions(+), 6 deletions(-)

diff --git a/northd/northd.c b/northd/northd.c
index 1ca8dedafa..2522df5b20 100644
--- a/northd/northd.c
+++ b/northd/northd.c
@@ -11881,3 +11881,6 @@ build_ecmp_routing_policy_flows(struct lflow_table *lflows,
-        ds_put_format(&actions, "%s = %"PRIu16
-                      "; %s = select(", REG_ECMP_GROUP_ID, ecmp_group_id,
-                      REG_ECMP_MEMBER_ID);
+        /* Hash on L4 ports like static-route selection_fields={tp_src,tp_dst}.
+         * Install protocol-specific flows so tcp/udp/sctp use their own
+         * port fields, and other IP traffic hashes on L3/proto only. */
+        struct ds proto_match = DS_EMPTY_INITIALIZER;
+        struct ds proto_actions = DS_EMPTY_INITIALIZER;
+        static const char *protocols[] = {NULL, "tcp", "udp", "sctp"};
@@ -11884,0 +11888,3 @@ build_ecmp_routing_policy_flows(struct lflow_table *lflows,
+        ds_put_format(&actions, "%s = %"PRIu16
+                      "; %s = select(values=(", REG_ECMP_GROUP_ID,
+                      ecmp_group_id, REG_ECMP_MEMBER_ID);
@@ -11892 +11898,28 @@ build_ecmp_routing_policy_flows(struct lflow_table *lflows,
-        ds_put_cstr(&actions, ");");
+
+        for (size_t i = 0; i < ARRAY_SIZE(protocols); i++) {
+            const char *protocol = protocols[i];
+
+            ds_clear(&proto_actions);
+            ds_put_cstr(&proto_actions, ds_cstr(&actions));
+            ds_clear(&proto_match);
+            if (protocol) {
+                ds_put_format(&proto_actions,
+                              "); hash_fields=\"ip_src,ip_dst,ipv6_src,"
+                              "ipv6_dst,ip_proto,%s_src,%s_dst\");",
+                              protocol, protocol);
+                ds_put_format(&proto_match, "(%s) && %s",
+                              rule->match, protocol);
+            } else {
+                ds_put_cstr(&proto_actions,
+                            "); hash_fields=\"ip_src,ip_dst,ipv6_src,"
+                            "ipv6_dst,ip_proto\");");
+                ds_put_format(&proto_match,
+                              "(%s) && (ip || ip6) && !tcp && !udp && !sctp",
+                              rule->match);
+            }
+            ovn_lflow_add(lflows, od, S_ROUTER_IN_POLICY, rule->priority,
+                          ds_cstr(&proto_match), ds_cstr(&proto_actions),
+                          lflow_ref, WITH_HINT(&rule->header_));
+        }
+        ds_destroy(&proto_match);
+        ds_destroy(&proto_actions);
@@ -11896,0 +11930,3 @@ build_ecmp_routing_policy_flows(struct lflow_table *lflows,
+        ovn_lflow_add(lflows, od, S_ROUTER_IN_POLICY, rule->priority,
+                      rule->match, ds_cstr(&actions), lflow_ref,
+                      WITH_HINT(&rule->header_));
@@ -11898,2 +11933,0 @@ build_ecmp_routing_policy_flows(struct lflow_table *lflows,
-    ovn_lflow_add(lflows, od, S_ROUTER_IN_POLICY, rule->priority, rule->match,
-                  ds_cstr(&actions), lflow_ref, WITH_HINT(&rule->header_));
